### ISSUE TITLE
Weak type detection doesn't consider generic types

### ISSUE THREAD

**Issue by sharwell** [Microsoft/TypeScript#16883]
**TypeScript Version:**  2.4.1

**Code**

```ts
interface BaseInterface {
  baseMethod?: () => void;
}

interface DerivedInterface extends BaseInterface {
  derivedMethod?: () => void;
}

class Impl implements DerivedInterface {
  derivedMethod(): void { }
}

class Helper {
  runCore<T extends BaseInterface>(o: T): void  { }

  run(): void {
    this.runCore(new Impl());
  }
}
```

**Expected behavior:**

TS2559 is not reported, since `Impl` shares a property `DerivedInterface`, an interface which extends `BaseInterface`.

**Actual behavior:**

TS2559 is reported.

**Workaround:**

If the following call is used, TS2559 is not reported:

```
this.runCore<DerivedInterface>(new Impl());
```

**Comment by sylvanaar** [comment]
The problem is that BaseInterface doesn't share a property with Impl(). This fix works too

```Typescript
class Helper {
  runCore<T extends BaseInterface>(o: T): void  { }

  run(): void {
    this.runCore(new Impl() as BaseInterface);
  }
}
`

**Comment by sharwell** [comment]
> The problem is that BaseInterface doesn't share a property with Impl().

It's not supposed to. The weak types implementation currently works well for certain algorithms, but is quite problematic for others. It could be substantially improved by avoiding the warning report when the object passed in shares a property with an *interface* that extends a weak type.

> This fix works too

The use of an unchecked cast precludes the approach from being considered a fix.

**Comment by BurtHarris** [comment]
Oh Sam, your are such a perfectionist.   It sounds like a fix to me, just not a clean fix.

**Comment by sandersn** [comment]
The problem is the structurally typed nature of typescript. @sylvanaar is right, from the compilers perspective, BaseInterface and Impl don't share any properties. This limitation is not limited to weak types; notice that you can't say `new Impl().baseMethod()` because the compiler doesn't even know it's there. The canonical workaround for this is interface merging:

```ts
interface Impl extends BaseInterface {}
class Impl implements DerivedInterface ...
```

This tells the compiler to add the base methods to Impl even if they aren't actually defined in the body of Impl. `implements` doesn't actually add anything to your type, it just checks that what is present conforms to the implemented interface. Afterward, the compiler has no idea that the class and the base interface are related. 

I think the concept BaseInterface really needs is "must implement at least one". Unfortunately, I don't think the OO concepts in typescript allow that. Maybe some clever union type would do it, though.
